### PR TITLE
Feature/machine state manager improv

### PR DESCRIPTION
**Pull request by ulianaStashkiv**

### CODE REVIEW

**Comment by adklimkiew** [comment]
There are too many changed files. Please describe which ones should be reviewed and what is the scope of this review.

**Comment by ulianaStashkiv** [comment]
State machine was added to Machine State Manager.

The list files that should be reviewed:
 1) execution manager:
        applications/ExecutionManager/include/execution_manager.hpp
        applications/ExecutionManager/src/execution_manager.cpp
 2) msm:
        applications/MachineStateManager/include/machine_state_manager.hpp
        applications/MachineStateManager/include/msm_state_machine.hpp
        applications/MachineStateManager/src/machine_state_manager.cpp
        applications/MachineStateManager/src/msm_main.cpp
        applications/MachineStateManager/src/msm_state_machine.cpp

But it would be better if you had time to review all changes.

File: applications/MachineStateManager/src/msm_state_machine.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    const char* applicationName = "MachineStateManager";
+    constexpr int defaulTimeout = 300000;
+
+    StateError result = m_msm.registerMsm(applicationName, defaulTimeout);
+
+    if(StateError::K_SUCCESS == result)
+    {
+        std::cout << "[ MachineStateManager ]:\tSuccessful registration as a MSM." << std::endl;
+        std::this_thread::sleep_for(std::chrono::seconds(5));
```
Comment (rBekesh) [inline applications/MachineStateManager/src/msm_state_machine.cpp]: magic number

File: applications/MachineStateManager/src/msm_state_machine.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        return std::make_unique<Shutdown>(m_msm);
+    }
+
+    return std::make_unique<Running>(m_msm);
+}
+
+void Running::enter()
+{
+    std::this_thread::sleep_for(std::chrono::seconds(5));
```
Comment (rBekesh) [inline applications/MachineStateManager/src/msm_state_machine.cpp]: magic number

File: applications/MachineStateManager/src/msm_state_machine.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    std::cout << "[ MachineStateManager ]:\tSetting machine state to RUNNING..." << std::endl;
+    m_msm.setMachineState("Running", 300000);
```
Comment (rBekesh) [inline applications/MachineStateManager/src/msm_state_machine.cpp]: magic number

File: applications/MachineStateManager/src/msm_state_machine.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    std::this_thread::sleep_for(std::chrono::seconds(5));
```
Comment (rBekesh) [inline applications/MachineStateManager/src/msm_state_machine.cpp]: magic number

File: applications/MachineStateManager/src/msm_state_machine.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    std::cout << "[ MachineStateManager ]:\tSetting machine state to LIVING..." << std::endl;
+    m_msm.setMachineState("Living", 300000);
```
Comment (rBekesh) [inline applications/MachineStateManager/src/msm_state_machine.cpp]: magic number

File: applications/MachineStateManager/src/msm_state_machine.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    std::this_thread::sleep_for(std::chrono::seconds(5));
```
Comment (rBekesh) [inline applications/MachineStateManager/src/msm_state_machine.cpp]: magic number

File: applications/MachineStateManager/src/msm_state_machine.cpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+    std::cout << "[ MachineStateManager ]:\tSetting machine state to SHUTTINGDOWN..." << std::endl;
+    m_msm.setMachineState("Shuttingdown", 300000);
```
Comment (rBekesh) [inline applications/MachineStateManager/src/msm_state_machine.cpp]: magic number

File: applications/ExecutionManager/src/em_main.cpp [hunk 8]
```diff
@@ -1,8 +1,10 @@
 #include "execution_manager_server.hpp"
 #include "execution_manager.hpp"
 #include "manifest_reader.hpp"
+
 #include <iostream>
 #include <memory>
+#include <unistd.h>
```
Comment (rBekesh) [inline applications/ExecutionManager/src/em_main.cpp]: not needed

File: applications/AdaptiveApplication1/src/state.cpp [hunk 9]
```diff
@@ -7,7 +7,8 @@ using ApplicationState = api::ApplicationStateClient::ApplicationState;
 State::State(AdaptiveApp &app, api::ApplicationStateClient::ApplicationState state, const std::string &stateName) :
     m_app{app}, m_applState{state}, m_stateName{stateName}
 {
-    std::cout << "Enter " << m_stateName << " state\n";
+//    std::cout << "[ AdaptiveApplication1 ]:\tEntering "
+//              << m_stateName << " styyyyyyyyyyyyyyyyyyyate." << std::endl;
```
Comment (rBekesh) [inline applications/AdaptiveApplication1/src/state.cpp]: remove commented code

File: applications/ExecutionManager/src/execution_manager.cpp [hunk 10]
```diff
@@ -242,20 +304,76 @@ ExecutionManager::setMachineState(pid_t processId, string state)
...
-            << "\"" << state << "\""
+  
+  std::cout << "[ ExecutionManager ]:\t\tInvalid machine state received - "
+            << "\"" << state << "\"."
             << std::endl;
 
   return false;
 }
 
+void ExecutionManager::confirmFromApplication()
```
Comment (adklimkiew) [inline applications/ExecutionManager/src/execution_manager.cpp]: Role of confirm is played by reportApplicationState I guess. Maybe rename to waitForReports or similar

File: applications/ExecutionManager/include/execution_manager.hpp [hunk 11]
```diff
@@ -123,7 +133,18 @@ class ExecutionManager
   std::vector<MachineState> m_machineManifestStates;
 
   std::string m_machineStateClientAppName;
+
   pid_t m_machineStateClientPid {-1};
+
+  std::map<pid_t, std::string> m_reportedApplications;
```
Comment (adklimkiew) [inline applications/ExecutionManager/include/execution_manager.hpp]: What is the purpose of this container?

File: applications/ExecutionManager/include/execution_manager.hpp [hunk 12]
```diff
@@ -123,7 +133,18 @@ class ExecutionManager
...
+
+  std::set<pid_t> m_stateConfirmToBeReceived;
+
+  void confirmFromApplication();
+  
+  std::set<std::string> getAllowedProcessForState(std::string);
```
Comment (adklimkiew) [inline applications/ExecutionManager/include/execution_manager.hpp]: What is the meaning of std::string here? appId, process name, machine state, other?

File: applications/ExecutionManager/src/execution_manager.cpp [hunk 13]
```diff
@@ -157,62 +167,79 @@ void ExecutionManager::startApplication(const ProcessInfo& process)
...
+              << "\" with pid "
+              << processId
+              << " started."
+              << std::endl;
   }
 
 }
 
 void
 ExecutionManager::reportApplicationState(pid_t processId, AppState state)
```
Comment (adklimkiew) [inline applications/ExecutionManager/src/execution_manager.cpp]: Does this call happen in a separate thread? If yes, then you need to protect shared resources. If no, how can you quit while loop in confirmFromApplication?

File: applications/ExecutionManager/src/execution_manager.cpp [hunk 14]
```diff
@@ -221,18 +248,53 @@ ExecutionManager::registerMachineStateClient(pid_t processId, string appName)
...
                                  state);
 
+  std::cout << "============================================================================" << std::endl;
+  std::cout << "[ ExecutionManager ]:\t\tSTATE TO BE SET : " << state<< std::endl;
+
+  std::set<std::string> allowedAppsSet = getAllowedProcessForState(state);
+  std::set<std::string> activeAppsSet = getActiveProcessForCurrentState();
+
+  std::set<std::string> processToBeKilledSet;
+  std::set<std::string> processToBeStartedSet;
```
Comment (adklimkiew) [inline applications/ExecutionManager/src/execution_manager.cpp]: Am I right to notice that this complex code and "heavy" data structures are used only for print/debug purposes??? Remove it

File: applications/ExecutionManager/src/execution_manager.cpp [hunk 15]
```diff
@@ -242,20 +304,76 @@ ExecutionManager::setMachineState(pid_t processId, string state)
...
+              << listOfPid
+              << std::endl;
+
+    kj::NEVER_DONE.wait(kj::setupAsyncIo().waitScope);
+
+    listOfPid.clear();      
+  }
+}
+
+std::set<std::string> ExecutionManager::getAllowedProcessForState(std::string state)
```
Comment (adklimkiew) [inline applications/ExecutionManager/src/execution_manager.cpp]: What is the purpose of this method? It basically copies data from member container and logs content. Remove it

File: applications/ExecutionManager/src/execution_manager.cpp [hunk 16]
```diff
@@ -242,20 +304,76 @@ ExecutionManager::setMachineState(pid_t processId, string state)
...
+    std::cout << id.processName << ", ";
+
+    allowedProc.insert(id.processName);
+  }
+  std::cout << std::endl;
+
+  return allowedProc;
+}
+
+std::set<std::string> ExecutionManager::getActiveProcessForCurrentState()
```
Comment (adklimkiew) [inline applications/ExecutionManager/src/execution_manager.cpp]: What is the purpose of this method? It basically copies data from member container and logs content. Remove it

File: applications/AdaptiveApplication1/src/state.cpp (hunk 9 above)
Comment (adklimkiew) [inline applications/AdaptiveApplication1/src/state.cpp]: What is this messy code?

File: applications/AdaptiveApplication2/src/adaptive_app.cpp [hunk 17]
```diff
@@ -0,0 +1,60 @@
+#include <adaptive_app.hpp>
+#include <state.hpp>
+
+#include <random>
+#include <iostream>
+
+using ApplicationState = api::ApplicationStateClient::ApplicationState;
+
```
Comment (adklimkiew) [inline applications/AdaptiveApplication2/src/adaptive_app.cpp]: Is this file copied entirely from Application1 ?? Why?

File: applications/AdaptiveApplication2/src/state.cpp [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+State::State(AdaptiveApp &app, api::ApplicationStateClient::ApplicationState state, const std::string &stateName) :
+    m_app{app}, m_applState{state}, m_stateName{stateName}
+{
+}
+
+void State::leave() const
+{
+
+}
+
```
Comment (adklimkiew) [inline applications/AdaptiveApplication2/src/state.cpp]: Is this file copied entirely from Application1 ?? Why?

File: applications/MachineStateManager/include/msm_state_machine.hpp [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    virtual void leave() const;
+
+protected:
+    api::ApplicationStateClient::ApplicationState getApplicationState() const;
+    MachineStateManager& m_msm;
+    const api::ApplicationStateClient::ApplicationState m_applState;
+    const std::string m_stateName;
+};
+
+class Init : public State
```
Comment (adklimkiew) [inline applications/MachineStateManager/include/msm_state_machine.hpp]: Again code copied from AdaptiveApp1

File: applications/ExecutionManager/include/execution_manager.hpp (hunk 11 above)
Comment (ulianaStashkiv) [reply applications/ExecutionManager/include/execution_manager.hpp]: This map is to track the applications that have reported their state.

File: applications/ExecutionManager/include/execution_manager.hpp (hunk 12 above)
Comment (ulianaStashkiv) [reply applications/ExecutionManager/include/execution_manager.hpp]: Process name

File: applications/MachineStateManager/include/msm_state_machine.hpp (hunk 19 above)
Comment (ulianaStashkiv) [reply applications/MachineStateManager/include/msm_state_machine.hpp]: In current implementation it was not possible to reuse state machine implementation from AdaptiveApp1 without deep refactoring of it so we decided to copy it.

**Comment by ulianaStashkiv** [comment]
Split to three branches.

**Pull request closed without merging**